File: archive/logic/TuringMachineSimulationNew/Tapes/CircularTape.cs
using TuringMachineSimulation.Logic.Tapes.TapesUtilities;

namespace TuringMachineSimulation.Logic.Tapes
{
    internal class CircularTape : ITape
    {
        public TapeTypes Type => TapeTypes.Circular;

        /// <summary>
        /// Stores the tape contents
        /// Blanks are not stored, boundaries of the tape represented by '>' and '<'
        /// </summary>
        public IDictionary<int, char> Tape { get; }
        
        /// <summary>
        /// Stores the scheduled write operations
        /// </summary>
        private Queue<WriteOperation> _writeQueue { get; }

        // Helper attributes for finding the position
        public int LeftBoundary { get; set; }
        public int RightBoundary { get; set; }

        /// <summary> CircularTape constructor, initializes the boundaries of the tape. </summary>
        /// <param name="leftBoundary"> The left boundary. Can be negative. </param>
        /// <param name="rightBoundary"> The right boundary. Can be negative. </param>
        public CircularTape(int leftBoundary = 0, int rightBoundary = 0)
        {
            Tape = new Dictionary<int, char>();
            _writeQueue = new Queue<WriteOperation>();
            
            LeftBoundary = leftBoundary;
            RightBoundary = rightBoundary;
            Tape[LeftBoundary - 1] = TapeSymbols.Start;
            Tape[RightBoundary + 1] = TapeSymbols.End;
        }
        
        public int GetMovedPosition(int position, int moves)
        {
            int tapeLength = RightBoundary - LeftBoundary + 1;

            int newPosition = position + moves;
            
            // Shift the range to [0, RightBoundary - LeftBoundary]
            // Obtain the new position, then shift back the original range
            newPosition = ((newPosition - LeftBoundary) % tapeLength + tapeLength) % tapeLength + LeftBoundary;
            
            return newPosition;
        }
        
        public bool IsOutOfRange(int position)
        {
            // circular tape will never be out of range
            return false;
        }

        public char Read(int position)
        {
            if (!Tape.TryGetValue(position, out var read))
                read = TapeSymbols.Blank;
            
            return read;
        }

        public bool TryRead(int position , out char? content)
        {
            content = !Tape.TryGetValue(position, out var read) ? TapeSymbols.Blank : read;
            return true;
        }

        public void ScheduleWrite(int position , char content, int machineID, int headID)
        {
            if (IsOutOfRange(position))
                throw new ArgumentOutOfRangeException();
            
            foreach (var op in _writeQueue.Where(op => op.Position == position))
            {
                throw new InvalidOperationException(
                    $"Machine '{machineID}' Head '{headID}' attempted to write '{content}', " + 
                    $"but Machine '{machineID}' Head '{op.HeadID}' has already scheduled a write to this position.");
            }
        
            var writeOperation = new WriteOperation(position, content, machineID, headID);
            _writeQueue.Enqueue(writeOperation);
        }

        public void CommitWrite()
        {
            while (_writeQueue.Count > 0)
            {
                WriteOperation writeOperation = _writeQueue.Dequeue();

                // Remove the key if the content written is a blank symbol
                if (writeOperation.Content == TapeSymbols.Blank)
                    Tape.Remove(writeOperation.Position);
                else
                    Tape[writeOperation.Position] = writeOperation.Content;
            }
        }
        
        public void InitializeContent(string contents)
        {
            if (contents.Length > RightBoundary - LeftBoundary + 1)
                throw new ArgumentOutOfRangeException(
                    "contents must be less than or equal to the tape length.");
            
            int i = 0;
            foreach (var content in contents)
            {
                Tape[i] = content;
                i++;
            }
        }
        
        public void DisplayContent(Dictionary<int, List<Tuple<int, int>>> headPositionsForTape, int tapeIndex)
        {
            int window = 10; // Display 10 cells

            Console.WriteLine($"Tape {tapeIndex}:");

            for (int i = -1; i < window; i++)
            {
                // Display the tape symbol
                if (Tape.TryGetValue(i, out var value))
                {
                    Console.Write($" {value} ");
                    if (value == TapeSymbols.End)
                        break;
                }
                else
                {
                    Console.Write($" {TapeSymbols.Blank} ");
                }

                // Display the head(s) at this position
                if (headPositionsForTape.TryGetValue(i, out var heads))
                {
                    Console.Write($"[{string.Join(",", heads.Select(tuple => $"M{tuple.Item1}H{tuple.Item2}"))}]");
                }

                Console.Write(" "); // Space between cells
            }

            Console.WriteLine(); // New line after displaying the tape
        }
        
        public void UpdateBoundaries(int headPosition)
        {
        }
        
        public string GetContentsAsString()
        {
            string contents = "";

            for (int i = LeftBoundary - 1; i <= RightBoundary + 1; i++)
            {
                if (Tape.TryGetValue(i, out var value))
                    contents += value;
                else
                    contents += TapeSymbols.Blank;
            }

            return contents;
        }
    }
}

File: frontend/turing-machine-web/src/logic/Tapes/CircularTape.ts
import { ITape } from "./ITape"
import { TapeTypes } from "./TapeTypes"
import { WriteOperation } from "./TapesUtilities/WriteOperation"
import { TapeSymbols } from "./TapesUtilities/TapeSymbols"
import { SignalState } from "../States/SignalStates";

export class CircularTape implements ITape
{
    // Getter for the tape type
    public get Type(): TapeTypes {
        return TapeTypes.Circular;
    }

    private _leftBoundary: number = 0;
    private _rightBoundary: number = 0;

    // Public getters for boundaries
    public get LeftBoundary(): number {
        return this._leftBoundary;
    }

    public get RightBoundary(): number {
        return this._rightBoundary;
    }

    private _tape: Map<number, string> = new Map();
    private _states: Map<number, SignalState> = new Map();
    private _writeQueue: WriteOperation[] = [];

    constructor(_leftBoundary: number = 0, _rightBoundary: number = 0)
    {
        this._leftBoundary = _leftBoundary;
        this._rightBoundary = _rightBoundary;

        // Stores the boundaries for displaying on the UI
        this._tape.set(this._leftBoundary - 1, TapeSymbols.Start)
        this._tape.set(this._rightBoundary + 1, TapeSymbols.End)
    }

    public IsOutOfRange(): boolean 
    {
        // Circular tape will never be out of range
        return false;
    }

    public Read(position: number): string 
    {
        if (this._tape.has(position))
            return TapeSymbols.Blank;

        return this._tape.get(position) || ' ';
    }

    public SendSignal(position: number): SignalState
    {
        if (!this._states.has(position))
            return SignalState.Other;

        return this._states.get(position)!;
    }

    public TryRead(position: number): { success: boolean; content: string | null } 
    {
        const read = this._tape.get(position);
        const content = read === undefined ? TapeSymbols.Blank : read;

        // Infinite tape will always be success
        return { success: true, content };
    }

    public ScheduleWrite(position: number, content: string, machineID: number, headID: number): void 
    {
        if (content.length != 1)
            throw new RangeError("The write content must be a single character.");
        
        for (let op of this._writeQueue)
        {
            if (op.Position == position)
            {
                throw new Error(
                    `Machine ${machineID} Head ${headID} attempted to write '${content}', ` +
                    `but Machine ${op.MachineID} Head ${op.HeadID} has already scheduled a write to this position.`
                );
            }
        }

        const writeOp = new WriteOperation(position, content, machineID, headID);
        this._writeQueue.push(writeOp);
    }

    public CommitWrite() 
    {
        // Poping the last element like in a stack
        // Since all operations happen at one step, a stack is enough here
        while (this._writeQueue.length > 0)
        {
            let op = this._writeQueue.pop()!;

            // Handle signal states
            if (op.Content == TapeSymbols.Pause || op.Content == TapeSymbols.Running) {
                this._states.set(op.Position, 
                    op.Content === TapeSymbols.Pause ? SignalState.Orange : SignalState.Green);
                
                continue;
            }
            
            // Handle regular tape content
            if (op.Content == TapeSymbols.Blank)
                this._tape.delete(op.Position);
            else
                this._tape.set(op.Position, op.Content);
        }
    }

    public RemoveSignal(position: number): void
    {
        this._states.delete(position);
    }

    public GetMovedPosition(position: number, moves: number): number {
        const tapeLength = this._rightBoundary - this._leftBoundary + 1;

        let newPosition = position + moves;
            
        // Shift the range to [0, rightBoundary - leftBoundary]
        // Obtain the new position, then shift back to the original range
        newPosition = ((newPosition - this._leftBoundary) % tapeLength + tapeLength) % tapeLength + this._leftBoundary;
            
        return newPosition;
    }

    public InitializeContent(contents: string): void 
    {
        for (let i = this._leftBoundary; i < this._leftBoundary + contents.length; i++)
        {
            this._tape.set(i, contents.charAt(i));
        }
    }

    public UpdateBoundaries(): void 
    {
        // The boundaries of Circular tape is fixed
    }

    public GetContentsAsString()
    {
        let contents = "";

        for (let i = this._leftBoundary - 1; i <= this._rightBoundary + 1; i++) {
            if (this._tape.has(i))
                contents += this._tape.get(i)!;
            else
                contents += TapeSymbols.Blank;
            
        }

        return contents;
    }
}